packages/central-server/app/patientPortal/auth/PortalOneTimeTokenService.js

```diff
@@ -0,0 +1,97 @@
+import { addMinutes } from 'date-fns';
+import { randomInt, randomBytes, createHash } from 'crypto';
+import { PORTAL_ONE_TIME_TOKEN_TYPES } from '@tamanu/constants';
+import { BadAuthenticationError } from '@tamanu/shared/errors';
+
+function randomSixDigitCode() {
+  // returns a zero-padded 6 digit string using crypto.randomInt for better security
+  return randomInt(100000, 1000000).toString();
+}
+
+function randomRegisterCode() {
+  // secure random code for magic links - 32 hex chars
+  return randomBytes(16).toString('hex');
+}
+
+function hashToken(token) {
+  return createHash('sha256').update(token).digest('hex');
```

Comment:
Yeah, no. We want a password-grade hash, not a data integrity hash. Argon2id or bcrypt.